nozzle.py

```diff
@@ -483,8 +482,20 @@ def __call__(self, x_vec, *, time=0, eos, **kwargs):
             )
             rho = pressure / temperature / gas_const
 
-            velocity = np.zeros(shape=(self._dim,))
-            velocity[self._direc] = self._mach*math.sqrt(gamma*pressure/rho)
+            @memoize_in(actx, ("set_velocity_knl", self._dim, self._direc))
+            def set_velocity_knl():
+                import loopy
+                from arraycontext import make_loopy_program
+                from arraycontext.transform_metadata import ElementwiseMapKernelTag
+                knl = make_loopy_program("{[i]: 0<=i<ndim}",
+                                         "vel[i] = v0 if i == direc else 0",
+                                         tags=(ElementwiseMapKernelTag(),))
+                return loopy.fix_parameters(knl, direc=self._direc,
+                                            ndim=self._dim)
+            velocity = actx.call_loopy(set_velocity_knl(),
+                                       v0=self._mach*actx.np.sqrt(gamma*pressure/rho)
+                                       )["vel"]
+            velocity = np.array(velocity)
```

Comment:
Is this actually necessary?  

This seems like an astounding amount of code around performing a simple arithmetic operation.  Consider that expressions just like, or very similar to, the original code being replaced here are littered throughout *MIRGE-Com*.  If this is what users need to do to perform arithmetic in *MIRGE-Com*, how will users ever write viable expressions?

Reply:
The problem isn't performing arithmetic, but performing an in-place update. We can expose an operation like `np.eye` to get the unit vector to get the velocity.

> how will users ever write viable expressions?

FWIW, there is arithmetic at other places in the driver which were left unchanged :).

Reply:
> The problem isn't performing arithmetic, but performing an in-place update. We can expose an operation like `np.eye` to get the unit vector to get the velocity.
> 

We do the type of thing done here in Mike's driver *a lot*, and I am just concerned that we would need to replace any of those with stuff like this.   What would be wrong with something like this then?

```python
Vx = <some calculation>
velocity = np.array([Vx, 0, 0])
```

> > how will users ever write viable expressions?
> 
> FWIW, there is arithmetic at other places in the driver which were left unchanged :).

Yeah, but that fact is cause for even more concern, because it seems that the user/developers have to basically guess what needs to be replaced, and also what to replace it with.

Reply:
Indeed there was an easier way to write it: 85a37a7 (Thanks!)

> Yeah, but that fact is cause for even more concern

The earlier code would continue to work with the `PyOpenCLArrayContext`. But that's a no-go in `PytatoArrayContext`, so if the earlier code-path is invoked with `PytatoArrayContext`, :mod:`pytato` will raise.

Reply:
Thanks for your continued patience with me, @kaushikcfd!  I would like to understand why your new expression is OK, but the old one was not.   This will help me avoid running afoul of such things going forward!

Reply:
By default, `numpy` chooses `dtype=float64` for storing the zeros. Assigning any elements of this array with a python object would result in an error.

In the lazy case, `Vx` would be a time-dependent entity => we would need to represent it symbolically i.e. `Vx` would be a thaw-ed array. In order to assign np arrays elements with Vx's symbolic expression we had to change the dtype of `zeros` to `object`.

Reply:
Ah, cool, this makes sense. The array is made into an object array and the entire expression for the non-zero component just drops right in instead of the number.  It would probably make sense for this particular quantity to be calculated using the original array context instead of the lazy one?  These are all constants set at init time.

Reply:
> It would probably make sense for this particular quantity to be calculated using the original array context instead of the lazy one?

There isn't much need to change the dtype for a `PyOpenCLArrayContext` as there would be no performance benefits for  the eager mode of computation (virtually no change in the #operations or #memory accesses). But it is important that we rewrite this for the benefit of lazy array-context. As, this quantity is a node in the DAG for which we would generate code later.

>These are all constants set at init time.

I think they are varying wrt time. `rho` is dependent on `pressure` that depends on `P0` which is a step function in terms of time set in `get_inflow_pressure`. Time is a symbolic input in our array expressions as we wouldn't want to repeatedly recompile at every time-step. Compilation is a pretty expensive operation relative to a single RHS evaluation.

Reply:
> > It would probably make sense for this particular quantity to be calculated using the original array context instead of the lazy one?
> > These are all constants set at init time.
> 
> I think they are varying wrt time. `rho` is dependent on `pressure` that depends on `P0` which is a step function in terms of time set in `get_inflow_pressure`. 

Oh yep, I missed that Mike was using a pressure ramp function here.